src/executor/top_n.rs

```diff
@@ -0,0 +1,160 @@
+                if heap.len() < heap_size {
+                    heap.push(row);
+                } else {
+                    let mut top = heap.peek_mut().unwrap();
+                    if cmp(&row, &top, &self.comparators) == Ordering::Less {
+                        *top = row;
+                    }
+                }
+            })
+        });
+
+        yield DataChunk::from_rows(
+            heap.into_sorted_vec()
+                .into_iter()
+                .skip(self.offset)
+                .take(self.limit)
+                .collect::<Vec<_>>()
+                .as_ref(),
+            &chunks[0],
+        );
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::ops::Range;
+
+    use futures::TryStreamExt;
+    use test_case::test_case;
+
+    use super::*;
+    use crate::array::ArrayImpl;
+    use crate::binder::{BoundExpr, BoundInputRef};
+    use crate::catalog::ColumnCatalog;
+    use crate::types::{DataTypeExt, DataTypeKind};
+
+    #[test_case(&[(0..6)], 1, 4, false, &[(1..5)])]
+    #[test_case(&[(0..6)], 0, 10, false, &[(0..6)])]
+    #[test_case(&[(0..6)], 10, 0, false, &[(0..0)])]
+    #[test_case(&[(0..2), (2..4), (4..6)], 1, 4, false, &[(1..5)])]
+    #[test_case(&[(0..6)], 1, 4, true, &[(1..5)])]
+    #[test_case(&[(0..6)], 0, 10, true, &[(0..6)])]
+    #[test_case(&[(0..6)], 10, 0, true, &[(0..0)])]
+    #[test_case(&[(0..2), (2..4), (4..6)], 1, 4, true, &[(1..5)])]
+    #[tokio::test]
+    async fn simple(
+        inputs: &'static [Range<i32>],
+        offset: usize,
+        limit: usize,
+        desc: bool,
+        outputs: &'static [Range<i32>],
+    ) {
+        let inputs: Vec<DataChunk> = inputs.iter().map(|r| range_to_chunk(false, r)).collect();
+
+        let (top_n, limit_order): (TopNExecutor, LimitExecutor) = equivalent_executors(
+            inputs,
+            offset,
+            limit,
+            vec![ColumnCatalog::new(
+                0,
+                DataTypeKind::Int(None).not_null().to_column("v1".into()),
+            )],
+            vec![(0, desc)],
+        );
+
+        let actual = top_n.execute().try_collect::<Vec<_>>().await.unwrap();
+        let outputs = outputs
+            .iter()
+            .map(|r| range_to_chunk(desc, r))
+            .collect_vec();
+        assert_eq!(actual, outputs);
+
+        let outputs_limit_order = limit_order.execute().try_collect::<Vec<_>>().await.unwrap();
+        assert_eq!(actual, outputs_limit_order);
+    }
+
+    fn range_to_chunk(reverse: bool, range: &Range<i32>) -> DataChunk {
+        let array = if reverse {
+            range.clone().rev().collect()
+        } else {
+            range.clone().collect()
+        };
+        [ArrayImpl::Int32(array)].into_iter().collect()
+    }
+
+    fn equivalent_executors(
+        inputs: Vec<DataChunk>,
+        offset: usize,
+        limit: usize,
+        catalog: Vec<ColumnCatalog>,
+        idx_desc: Vec<(usize, bool)>,
+    ) -> (TopNExecutor, LimitExecutor) {
```

Comment:
Actually I'm considering randomized/property-based testing.🤔 (Maybe we can do it in the future)